## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'
application_df = application_df.drop(columns=['EIN', 'NAME','USE_CASE'])
application_df.head(5)

APPLICATION_TYPE       AFFILIATION CLASSIFICATION  ORGANIZATION  STATUS  \
0              T10       Independent          C1000   Association       1   
1               T3       Independent          C2000  Co-operative       1   
2               T5  CompanySponsored          C3000   Association       1   
3               T3  CompanySponsored          C2000         Trust       1   
4               T3       Independent          C1000         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [3]:
# Determine the number of unique values in each column.
unique_counts = application_df.nunique()
print(unique_counts)


APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64


In [4]:
# Look at APPLICATION_TYPE value counts for binning
application_type_counts = application_df['APPLICATION_TYPE'].value_counts()
print(application_type_counts)

APPLICATION_TYPE
T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: count, dtype: int64


In [5]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`

cutoff_value = 1500
types = application_type_counts[application_type_counts < cutoff_value].index.tolist()
application_types_to_replace = types

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

APPLICATION_TYPE
T3       27037
Other     5720
T4        1542
Name: count, dtype: int64

In [6]:
# Look at CLASSIFICATION value counts for binning
bin_count = application_df.CLASSIFICATION.value_counts()
bin_count

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C1248        1
C6100        1
C1820        1
C1900        1
C2150        1
Name: count, Length: 71, dtype: int64

In [7]:
# You may find it helpful to look at CLASSIFICATION value counts >1
bin_count[bin_count> 1]

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C2300       32
C7200       32
C1240       30
C8000       20
C7120       18
C1500       16
C6000       15
C1800       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C1720        6
C4100        6
C2400        6
C1600        5
C1257        5
C2710        3
C0           3
C1260        3
C1256        2
C1234        2
C1267        2
C1246        2
C3200        2
Name: count, dtype: int64

In [8]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
# Choose a cutoff value for binning
classification_cutoff_value = 2000

classification_counts = application_df['CLASSIFICATION'].value_counts()
classifications_to_replace = classification_counts[classification_counts < classification_cutoff_value].index.tolist()

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

CLASSIFICATION
C1000    17326
C2000     6074
Other     6062
C1200     4837
Name: count, dtype: int64

In [9]:
# Convert categorical data to numeric with `pd.get_dummies`
application_dummies = pd.get_dummies(application_df)
application_dummies

STATUS   ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0           1      5000              1                    True   
1           1    108590              1                   False   
2           1      5000              0                    True   
3           1      6692              1                   False   
4           1    142590              1                   False   
...       ...       ...            ...                     ...   
34294       1      5000              0                   False   
34295       1      5000              0                   False   
34296       1      5000              0                   False   
34297       1      5000              1                    True   
34298       1  36500179              0                   False   

       APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  AFFILIATION_CompanySponsored  \
0                    False                False                         False   
1                     True                False                         False   
2                    False                False                          True   
3                     True                False                          True   
4                     True                False                         False   
...                    ...                  ...                           ...   
34294                False                 True                         False   
34295                False                 True                          True   
34296                 True                False                          True   
34297                False                False                         False   
34298                 True                False                         False   

       AFFILIATION_Family/Parent  AFFILIATION_Independent  \
0                          False                     True   
1                          False                     True   
2                          False                    False   
3                          False                    False   
4                          False                     True   
...                          ...                      ...   
34294                      False                     True   
34295                      False                    False   
34296                      False                    False   
34297                      False                     True   
34298                      False                     True   

       AFFILIATION_National  ...  INCOME_AMT_1-9999  INCOME_AMT_10000-24999  \
0                     False  ...              False                   False   
1                     False  ...               True                   False   
2                     False  ...              False                   False   
3                     False  ...              False                    True   
4                     False  ...              False                   False   
...                     ...  ...                ...                     ...   
34294                 False  ...              False                   False   
34295                 False  ...              False                   False   
34296                 False  ...              False                   False   
34297                 False  ...              False                   False   
34298                 False  ...              False                   False   

       INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  \
0                         False               False             False   
1                         False               False             False   
2                         False               False             False   
3                         False               False             False   
4                          True               False             False   
...                         ...                 ...               ...   
34294                     False               Fa

In [10]:
# Split our preprocessed data into our features and target arrays
X = application_dummies.drop(["IS_SUCCESSFUL"], axis='columns').values
y = application_dummies["IS_SUCCESSFUL"].values


# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [11]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

**Optimization Attempt no:1**
- Dropping fewer columns
- Creating more bins for rare occurences in columns

## Compile, Train and Evaluate the Model

In [12]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len(X_train[0])
hidden_nodes_layer1 =  80
hidden_nodes_layer2 = 30


nn1 = tf.keras.models.Sequential()

# First hidden layer
nn1.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu"))


# Second hidden layer
nn1.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Output layer
nn1.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn1.summary()


# Compile the model
nn1.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

# Train the model
fit_model = nn1.fit(X_train_scaled, y_train, epochs=100)

# Evaluate the model using the test data
model_loss, model_accuracy = nn1.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 80)                2480      
                                                                 
 dense_1 (Dense)             (None, 30)                2430      
                                                                 
 dense_2 (Dense)             (None, 1)                 31        
                                                                 
Total params: 4941 (19.30 KB)
Trainable params: 4941 (19.30 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
Epoch 1/100
804/804 [==============================] - 6s 3ms/step - loss: 0.5920 - accuracy: 0.7074
Epoch 2/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5793 - accuracy: 0.7142
Epoch 3/100
804/804 [==============================] - 5s 6ms/step - loss: 0.5770 - 

**Optimization Attempt No:2**
- Add more neurons to a hidden layer.
- Add more hidden layers.

In [14]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len(X_train[0])
hidden_nodes_layer1 =  100
hidden_nodes_layer2 = 50
hidden_nodes_layer3 = 30


nn2 = tf.keras.models.Sequential()

# First hidden layer
nn2.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu"))


# Second hidden layer
nn2.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Third hidden layer
nn2.add(tf.keras.layers.Dense(units=hidden_nodes_layer3, activation="relu"))

# Output layer
nn2.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn2.summary()


# Compile the model
nn2.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

# Train the model
fit_model = nn2.fit(X_train_scaled, y_train, epochs=100)

# Evaluate the model using the test data
model_loss, model_accuracy = nn2.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_7 (Dense)             (None, 100)               3100      
                                                                 
 dense_8 (Dense)             (None, 50)                5050      
                                                                 
 dense_9 (Dense)             (None, 30)                1530      
                                                                 
 dense_10 (Dense)            (None, 1)                 31        
                                                                 
Total params: 9711 (37.93 KB)
Trainable params: 9711 (37.93 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
Epoch 1/100
804/804 [==============================] - 6s 5ms/step - loss: 0.5876 - accuracy: 0.7091
Epoch 2/100
804/804 [==============================

**Optimization Attempt No:3**
- Use different activation functions for the hidden layers.
- Add or reduce the number of epochs to the training regimen.

In [15]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len(X_train[0])
hidden_nodes_layer1 =  80
hidden_nodes_layer2 = 30


nn3 = tf.keras.models.Sequential()

# First hidden layer
nn3.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="tanh"))


# Second hidden layer
nn3.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="tanh"))

# Output layer
nn3.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn3.summary()


# Compile the model
nn3.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

# Train the model
fit_model = nn3.fit(X_train_scaled, y_train, epochs=150)

# Evaluate the model using the test data
model_loss, model_accuracy = nn3.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_11 (Dense)            (None, 80)                2480      
                                                                 
 dense_12 (Dense)            (None, 30)                2430      
                                                                 
 dense_13 (Dense)            (None, 1)                 31        
                                                                 
Total params: 4941 (19.30 KB)
Trainable params: 4941 (19.30 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
Epoch 1/150
804/804 [==============================] - 4s 3ms/step - loss: 0.5916 - accuracy: 0.7080
Epoch 2/150
804/804 [==============================] - 3s 4ms/step - loss: 0.5775 - accuracy: 0.7148
Epoch 3/150
804/804 [==============================] - 4s 5ms/step - loss: 0.5751 

In [16]:
# Export our model to HDF5 file
nn2.save("AlphabetSoupCharity_Optimization.h5")

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


##Overview of the analysis## 
The purpose of this analysis to develop a machine learning techniques, particularly neural networks, to predict whether applicants for funding from Alphabet Soup will be successful in their ventures. This classifier will analyze the provided dataset, which includes information on over 34,000 organizations that have received funding from Alphabet Soup. The dataset contains various features such as application type, affiliation, classification, use case, organization type, income amount, special considerations, funding amount requested, and whether the funding was used effectively.




##RESULTS##
•	Data Preprocessing
1.	What variable(s) are considered the target for your model?
The target variable for this model is the "Is-Successful" column. 

2.	What variable(s) are considered to be the features for your model?
APPLICATION_TYPE, AFFILIATION, CLASSIFICATION, ORGANIZATION, INCOME_AMT, SPECIAL_CONSIDERATIONS, STATUS, and ASK_AMT are the features variable for this model.

3.	What variable(s) are neither and should be removed from the input data? 
Columns like EIN (Employer Identification Number) and NAME was dropped because they are unique identifiers and  NAME column had too many unique numbers for predictive modeling. USE_CASE column provide very little information for prediction and can be removed.


•	Compiling, Training, and Evaluating the Model
1.	How many neurons, layers, and activation functions did you select for your neural network model, and why?
In this model I made 3 attempts to increase the accuracy above 75%.
In attempt one -   There are two hidden layers each with 80 and 30 neurons. The first and second activation function was 'relu' and the output function was 'sigmoid'. 

In attempt two –There are three hidden layers each with 100, 50 and 30 neurons each. The first, second, and third activation function was 'relu' and the output function was 'sigmoid'. 

In attempt three – There are two hidden layers each with 80 and 30 neurons. The first and second activation function was 'tanh' and the output function was 'sigmoid'. 


2.	Were you able to achieve the target model performance?
No


3.	What steps did you take to try and increase model performance?
In this model I made 3 attempts to increase the accuracy above 75%.
In attempt one -   I tried dropping three columns EIN, NAME and USE_CASE and creating more bins for rare occurrences in column APPLICATION_TYPE with cut off value of 1500 and  column CLASSIFICATION with cut off value of 2000. I also increased the number of values for each bin.  There are two hidden layers each with 80 and 30 neurons. The first and second activation function was 'relu' and the output function was 'sigmoid'. 

In attempt two – I tried to add more neurons to hidden layers and  also add more hidden layers. There are three hidden layers each with 100, 50 and 30 neurons each. The first,second, and third activation function was 'relu' and the output function was 'sigmoid'. 

In attempt three – I tried to use different activation functions for the hidden layers and increased the number of epochs to the training regimen. There are two hidden layers each with 80 and 30 neurons. The first and second activation function was 'tanh' and the output function was 'sigmoid'. The number of epochs was increased from 100 to 150.

#### Summary####
Overall, I was not able to achieve the accuracy above 75%. However I tried three different methods to achieve that. Out of the three attempts, attempt 2 was successful in gaining accuracy of 0.7140. Attempt one had accuracy of 0.7132 and attempt three had accuracy of 0.7137. And, an applicant has a 71% chance of being successful if they have the following:
- The type of APPLICATION is one of the following; T3, T4, and other.
- The application has the following CLASSIFICATION; C1000, C2000 and C1200.

A good model to recommend is the Random Forest model or Decision tree model because they are good for classification problems. 
